### PR TITLE
Handle templated object function parameters

### PR DESCRIPTION
**Pull request by weirdan**
Fixes vimeo/psalm#3397

### CODE REVIEW

**Comment by weirdan** [comment]
There are tons of psalm issues currently reported in master, I'd prefer to rebase this after they are resolved.

File: src/Psalm/Internal/Analyzer/Statements/Expression/Assignment/PropertyAssignmentAnalyzer.php [hunk 1]
```diff
@@ -237,6 +238,20 @@ public static function analyzeInstance(
                     continue;
                 }
 
+                if ($lhs_type_part instanceof Type\Atomic\TTemplateParam) {
+                    $extra_types = $lhs_type_part->extra_types;
+
+                    $lhs_type_part = array_values(
+                        $lhs_type_part->as->getAtomicTypes()
+                    )[0];
```
Comment (afbarbaro) [inline src/Psalm/Internal/Analyzer/Statements/Expression/Assignment/PropertyAssignmentAnalyzer.php:246]: Is this guaranteed to always have at lest one element?

File: src/Psalm/Internal/Analyzer/Statements/Expression/Assignment/PropertyAssignmentAnalyzer.php [hunk 2]
```diff
@@ -789,7 +804,7 @@ public static function analyzeInstance(
                     if (IssueBuffer::accepts(
                         new InvalidPropertyAssignment(
                             $lhs_var_id . ' with non-object type \'' . $invalid_assignment_type .
-                            '\' cannot treated as an object',
+                            '\' cannot be treated as an object',
```
Comment (afbarbaro) [inline src/Psalm/Internal/Analyzer/Statements/Expression/Assignment/PropertyAssignmentAnalyzer.php:807]: 👍

File: src/Psalm/Internal/Analyzer/Statements/Expression/Assignment/PropertyAssignmentAnalyzer.php [hunk 3]
```diff
@@ -237,6 +238,20 @@ public static function analyzeInstance(
                     continue;
                 }
 
+                if ($lhs_type_part instanceof Type\Atomic\TTemplateParam) {
+                    $extra_types = $lhs_type_part->extra_types;
```
Comment (afbarbaro) [inline src/Psalm/Internal/Analyzer/Statements/Expression/Assignment/PropertyAssignmentAnalyzer.php:242]: Do we need to assign this to a variable here? It seems it's only needed if we enter into the if on lines 250-251 and it just gets re-assigned.  Maybe able to do without it here.

**Comment by muglug** [comment]
Should be safe to rebase now, sorry.

I'm mid-cleanup, but I promise the next round won't be so messy.

File: src/Psalm/Internal/Analyzer/Statements/Expression/Assignment/PropertyAssignmentAnalyzer.php (hunk 1 above)
Comment (weirdan) [reply src/Psalm/Internal/Analyzer/Statements/Expression/Assignment/PropertyAssignmentAnalyzer.php:246]: Yes. `getAtomicTypes()` returns `non-empty-array`, `array_values()` turn that into `non-empty-list`, so there's always an element at offset `0`.

File: src/Psalm/Internal/Analyzer/Statements/Expression/Assignment/PropertyAssignmentAnalyzer.php (hunk 3 above)
Comment (weirdan) [reply src/Psalm/Internal/Analyzer/Statements/Expression/Assignment/PropertyAssignmentAnalyzer.php:242]: `$lhs_type_part` is overwritten on the line below. So `$extra_types` is reassigned, but to the property of a different object.

**Comment by muglug** [comment]
I fixed this _slightly_ differently in f335560

**Pull request closed without merging**